### PR TITLE
@pepopowitz => Emit google ad snippet, and make sure it is rendered + evaluated once

### PR DESCRIPTION
**Pull request by mzikherman**
Much thanks to @zephraph for pairing and helping to identify and fix an underlying bug in our app code that was causing `<head>` tags to accumulate and be duplicated across two different render passes.

This PR includes that bug fix as well as emitting a Google retargeting script, only for artworks that are acquirable or in an auction. Locally this seems to work and generate the right events (cc @cavvia ).

That ticket is https://artsyproduct.atlassian.net/secure/RapidBoard.jspa?rapidView=57&projectKey=DISCO&modal=detail&selectedIssue=DISCO-962

This also 🤞 hopefully fixes https://artsyproduct.atlassian.net/secure/RapidBoard.jspa?rapidView=57&projectKey=DISCO&modal=detail&selectedIssue=DISCO-875 , since with the underlying bug fix that data is only rendered once.

### CODE REVIEW

File: src/Apps/Artwork/Components/ArtworkMeta.tsx [hunk 1]
```diff
@@ -71,6 +73,47 @@ export class ArtworkMeta extends Component<ArtworkMetaProps> {
     }
   }
 
+  renderGoogleAdSnippet() {
+    const { artwork, googleAdId: fromPropsGoogleAdId } = this.props
+    const { GOOGLE_ADWORDS_ID: fromSharifyGoogleAdId } = sd
+    const { is_in_auction, is_acquireable, _id } = artwork
+    if (!is_in_auction && !is_acquireable) return
+
+    // TODO: Investigate always being able to select from sharify.
```
Comment (mzikherman) [inline src/Apps/Artwork/Components/ArtworkMeta.tsx]: This dance is b/c during the SSR render, `sd.GOOGLE_ADWORDS_ID` is undefined. So we wind up passing that in to the app context from Force  via `res.locals.sd.GOOGLE_ADWORDS_ID`.

I think this is what we have to do to be able to consistently access that value properly on both the server and the client (this follows a pattern we've used elsewhere).

This should be safely mergeable in the meantime, and there's an accompanying small Force PR.

File: src/Artsy/Router/buildServerApp.tsx [hunk 2]
```diff
@@ -105,7 +105,7 @@ export function buildServerApp(config: ServerRouterConfig): Promise<Resolve> {
             const data = await relayEnvironment.relaySSRMiddleware.getCache()
             // Render tree again, but this time with Relay data being available.
             const html = ReactDOMServer.renderToString(
```
Comment (mzikherman) [inline src/Artsy/Router/buildServerApp.tsx]: This was the bug fix, if you expand the lines here you'll see that `ReactDOMServer.renderToString` is called twice, and accumulates `headTags` through both passes. The bug fix here was to only ever pass those in and accumulate them once.

**Comment by artsyit** [comment]
Deploy preview for *artsy-reaction* ready!

Built with commit 4730d751a2948bd32cff9db159ae64794bb867c2

https://deploy-preview-2327--artsy-reaction.netlify.com

File: src/Artsy/Router/buildServerApp.tsx (hunk 2 above)
Comment (pepopowitz) [reply src/Artsy/Router/buildServerApp.tsx]: I'm having trouble understanding how `headTags` is accumulating across those two calls. Is it being mutated by the `<Boot>` component? The only place I see it being set or changing is https://github.com/artsy/reaction/pull/2327/files#diff-8721a7de273dd9c3f5ca4c9838197b6eL79.

Comment (mzikherman) [reply src/Artsy/Router/buildServerApp.tsx]: I'm a tad fuzzy on that myself, but yes, this winds up being referenced inside `Boot`: https://github.com/artsy/reaction/blob/0d3c391bbc4233bdf96175ed144626e9f2856a5e/src/Artsy/Router/Boot.tsx#L58 

If looking at `HeadProvider`, https://github.com/tizmagik/react-head/blob/master/src/HeadProvider.js , there is indeed some `const var = this.props.headTags; var.push(...);`, etc. but I didn't trace it further.

When pairing we had a simple repro that was being popping up twice `alert("test")`, and with this fix went down to once!

Comment (mzikherman) [reply src/Artsy/Router/buildServerApp.tsx]: Also, if you view source, you'll see the tags all being duplicated. And with this fix the HTML looks correct (no dupes).

Comment (zephraph) [reply src/Artsy/Router/buildServerApp.tsx]: It is being mutated by `Boot`, yes. 

`headTags` is an array that gets passed to the `HeadProvider`. When you use the components from `react-head`, they're actually context consumers which have a reference to that array. They basically just push stuff onto it and the ultimate rendering in the head via the portal is handled at a higher level. 

We render twice. Once for data fetching and once to actually render the view. `headTags` was being passed down both times and thus being duplicated.

Comment (pepopowitz) [reply src/Artsy/Router/buildServerApp.tsx]: Wow, I wouldn't have expected mutation of props. I guess it makes sense for what `react-head` is doing, it's just something I totally take for granted with my strict personal "never ever mutate props" rule. (And maybe this is just another lesson to myself to never ever say never ever.)

On that note, one suggestion I'd make is that we name `headTags` something that makes it clear that it's going to be mutated, or add a comment indicating such? Definitely not required for this PR, but possibly helpful for future debuggers.

**Comment by pepopowitz** [comment]
This is really great, thank you both _sooooooo much_ for digging into this. 

I have only my comment about [giving hints to future readers that `headTags` is mutated down-stream](https://github.com/artsy/reaction/pull/2327#discussion_r277889243). If you feel like making that change, ❤️. If not, feel free to merge!

File: src/Artsy/Router/buildServerApp.tsx [hunk 3]
```diff
@@ -79,12 +79,12 @@ export function buildServerApp(config: ServerRouterConfig): Promise<Resolve> {
         const headTags = [<style type="text/css">{MediaStyle}</style>]
```
Comment (zephraph) [inline src/Artsy/Router/buildServerApp.tsx]:
```suggestion
        /**
         * An array that gets passed to `react-head`'s provider that will collect the header
         * tags that are rendered by the App. `headTags` is _mutatated_ when it's passed to the App. Beware.
         **/
        const headTags = [<style type="text/css">{MediaStyle}</style>]
```

Comment (zephraph) [reply src/Artsy/Router/buildServerApp.tsx]: cc @pepopowitz

Comment (mzikherman) [reply src/Artsy/Router/buildServerApp.tsx]: I couldn't accept the change for some reason :(

Applied in a commit.

**Comment by artsyit** [comment]
:rocket: PR was released in v16.3.3 :rocket:

**Comment by cavvia** [comment]
Thanks for this @mzikherman and @zephraph! I've QA'd on staging to verify the script is included on the correct artwork pages. The call has the correct pixel id and correct `dynx_itemid` so looks good to me. :+1:

**Pull request merged**